main/src/library/BigInt.flix

```diff
@@ -97,6 +97,116 @@ namespace BigInt {
+    ///
+    /// If the numeric value of `x` can be represented exactly return `Some(x as Int8)`.
+    ///
+    /// If the numeric value of `x` is outside the range of Int8 return `None`.
+    ///
+    pub def toInt8(x: BigInt): Option[Int8] =
+        import java.math.BigInteger.byteValueExact();
+        try {
+            Some(byteValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Int16]`.
+    ///
+    /// If the numeric value of `x` can be represented exactly return `Some(x as Int16)`.
+    ///
+    /// If the numeric value of `x` is outside the range of Int16 return `None`.
+    ///
+    pub def toInt16(x: BigInt): Option[Int16] =
+        import java.math.BigInteger.shortValueExact();
+        try {
+            Some(shortValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Int32]`.
+    ///
+    /// If the numeric value of `x` can be represented exactly return `Some(x as Int32)`.
+    ///
+    /// If the numeric value of `x` is outside the range of Int32 return `None`.
+    ///
+    pub def toInt32(x: BigInt): Option[Int32] =
+        import java.math.BigInteger.intValueExact();
+        try {
+            Some(intValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Int64]`.
+    ///
+    /// If the numeric value of `x` can be represented exactly return `Some(x as Int64)`.
+    ///
+    /// If the numeric value of `x` is outside the range of Int64 return `None`.
+    ///
+    pub def toInt64(x: BigInt): Option[Int64] =
+        import java.math.BigInteger.longValueExact();
+        try {
+            Some(longValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Float32]`.
+    ///
+    /// If the numeric value of `x` is within the range of Float32 return `Some(x as Float32)`.
+    ///
+    /// Warning: even if `x` is in the range of Float32 it may lose precision.
+    ///
+    /// If the numeric value of `x` is outside the range of Float32 return `None`.
+    ///
+    pub def toFloat32(x: BigInt): Option[Float32] =
+        use Option.flatMap;
+        import new java.math.BigDecimal(BigInt) as newBigDecimal;
+        import java.math.BigDecimal.floatValue();
+        let xd = newBigDecimal(x) as & Pure;
+        let d = xd.floatValue() as & Pure;
+        if (Float32.isInfinite(d) || Float32.isNan(d))
+            None
+        else
+            let* f32min = Float32.toBigInt(Float32.minValue());
+            let* f32max = Float32.toBigInt(Float32.maxValue());
+            if (x < f32min || x > f32max)
+                None
+            else
+                Some(d)
```

Comment:
I think I understand the last part. You ensure that the number is not too big or small to fit within a float. What does the first part do? How can the number become infinite or NaN?  Looking at the JavaDoc for floatValue I see that it can return infinity or negative infinity. I guess that `isInfinite` checks for both. But what about NaN? 

But wait, if floatValue() can return infinity, why do we need the latter checks? Is something more subtle going on?

Reply:
Side note: I guess we should add support for BigDecimal at some point.

Reply:
The inifinity checks are redundant - because d was originally a BigInt it would not get converted to either -Infinity or +Infinity. I've pushed a change to fix this.

BigDecimal support would be good.